File: 328_OddEvenLinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        if (head == null || head.next == null || head.next.next == null) {
            return head;
        }

        ListNode odd = head;
        ListNode even = head.next;
        ListNode t1 = head;
        ListNode t2 = head.next;

        while (t1 != null && t2 != null && t1.next != null && t2.next != null) {
            t1.next = t1.next.next;
            t2.next = t2.next.next;
            t1 = t1.next;
            t2 = t2.next;
        }
        t1.next = even;
        return odd;
    }
}

File: c++/328_OddEvenLinkedList.cpp
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        auto pH = new ListNode(-1), qH = new ListNode(-1);
        auto p = pH, q = qH, cur = head;
        bool flag = true;
        while (cur) {
            if (flag) {
                p = p->next = cur;
            } else {
                q = q->next = cur;
            }
            flag = !flag;
            cur = cur->next;
        }
        p->next = qH->next;
        q->next = nullptr;
        return pH->next;
    }
};